XML behavior with parameters with children but without group type

Ashwolaa opened the pull request:
Parameters in pyqtgraph can have children without the necessity for them to be of group type.

This behavior is currently not being adressed as the type of parameters define how it will be explored and written.

With this patch, it is now possible to correctly write and read an XML file which uses parameter with:
1) non-group type
2) children

Ashwolaa committed:
Correctly write and read an XML file which uses parameter with:
1) non-group type
2) children

Ashwolaa committed:
Merge branch 'PyMoDAQ:5.0.x_dev' into features/params_group_type

Ashwolaa commented:
Wrong branch

seb5g commented:
could you add such a parameter in the parameter_ex? We use that guy for all the testing so your stuff would be tested also!

Ashwolaa committed:
Merge branch 'PyMoDAQ:5.0.x_dev' into features/params_group_type

Ashwolaa committed:
Updating ParameterEx to provide example

Ashwolaa committed:
Adding an assert value in the testing

Ashwolaa commented:
> could you add such a parameter in the parameter_ex? We use that guy for all the testing so your stuff would be tested also!

This is in place, I added it in the group category

seb5g commented:
actually are you using this functionnality? I'm a bit worried it would create issues in some things I programmed assuming only a group would have children...

seb5g commented:
we should therefore be particularly cautious with previous filtering using the 'group' string in type. 
For instance, in ioxml line 45:
```
if 'group' not in param_type:  # covers 'group', custom 'groupmove', 'groupai' ...
    add_text_to_elt(parent_elt, param)
```

this could trigger an issue. add_text_to_elt should be used when there is a value to be written as text in the xml element. Therefore the  if clause should be:
```
if 'value' in param.opts:
```
I know this one is not directly related to your PR but we should properly review all the locations of such use of the check for 'group' in param.type() kind of clause

Ashwolaa commented:
> actually are you using this functionnality? I'm a bit worried it would create issues in some things I programmed assuming only a group would have children...

I am using it in my roi_dev branch. In general, I have always been using this functionnality within my GUIs. It is especially useful for instruments with many channels. The checkbox parameter used as a group provides a neat way to enable or disable channels.

Ashwolaa commented:
> we should therefore be particularly cautious with previous filtering using the 'group' string in type. For instance, in ioxml line 45:
> 
> ```
> if 'group' not in param_type:  # covers 'group', custom 'groupmove', 'groupai' ...
>     add_text_to_elt(parent_elt, param)
> ```
> 
> this could trigger an issue. add_text_to_elt should be used when there is a value to be written as text in the xml element. Therefore the if clause should be:
> 
> ```
> if 'value' in param.opts:
> ```
> 
> I know this one is not directly related to your PR but we should properly review all the locations of such use of the check for 'group' in param.type() kind of clause

I agree with you, I wanted to do minimal changes to avoid problems with the previous implementation but I'd much rather change 
 it as you suggest. I will look through the codes to check if I find other cases like this one.

seb5g commented:
thx for looking into this. My student is actually working on implementing a factory pattern to create xml and back from string. I think we can treat the value as any other "opts" and make the whole thing a bit simpler. But we will start our modifications from this PR once its ready @PauloGitHB I'm talking about your work ;-)

Ashwolaa committed:
- adding label keyword in dict_from_param (used by bool_push)
- using 'value' in param.opts for criteria to access add_text_to_elt

Ashwolaa committed:
adding label keyword in elt_to_dict

Ashwolaa committed:
Some formatting and cleaning up redundancies

Ashwolaa commented:
I agree, there is actually no reason to not include an extra entry with keyword value. You could say that it is redundant with the text attribute but it gets much clearer in the management. It should be quite straightforward to adapt the functions in the ioxml file.

I think it is good for a go, the tests are all getting passed and the behavior with loading/saving/updating in the parameter manager is ok too.

Btw, I realize we were not handling well the label in the bool_push parameter item.
commit 2f4533ae7b1c44276f76ec28defce6c91dbbee23 and commit a053153948baa991fda66f87601e61b2d1f271f8 fixes it.

Ashwolaa committed:
Adding keyworkds to filter children parameters according to their name or their types

Ashwolaa committed:
- Adding possibility to select filtered types
- Adding tests

Ashwolaa committed:
now works at intended

Ashwolaa commented:
I added in this PR some small changes in the iter_children_params function (parameter/utils.py) that allows one to filter parameters from the list.
For now, I added a filter for the 1) type or for the 2) name 
The user can decide if he wants to keep or not keep the filtered parameters with a boolean argument. 
For example, this can be convenient to remove the group type which does not have any value or also to select some specific parameter types or names

Without any inputs, the behavior is the same as before.

I added a test (test_iter_children_params) in param_utils_test.py checking that the behavior is consistent

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -166,14 +166,33 @@ def iter_children(param, childlist=[]):
 
 
-def iter_children_params(param, childlist=[]):
+def iter_children_params(param, childlist=[], filter_type=[], filter_name=[], selectFilter=False):
```
use tuples that are immutable for your filters, and use lower case for arguments

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -166,14 +166,33 @@ def iter_children(param, childlist=[]):
-def iter_children_params(param, childlist=[]):
+def iter_children_params(param, childlist=[], filter_type=[], filter_name=[], selectFilter=False):
     """Get a list of parameters under a given Parameter
+        | Returns all children parameters.
```
could you reformat the whole docstring using numpy style (the one here is an old attempt to follow pyqtgraph style but it has been abandoned)

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -166,14 +166,33 @@ def iter_children(param, childlist=[]):
     """
     for child in param.children():
-        childlist.append(child)
+        isFiltered = child.type() in filter_type or child.name() in filter_name
```
lower case!

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -166,14 +166,33 @@ def iter_children(param, childlist=[]):
     for child in param.children():
-        childlist.append(child)
+        isFiltered = child.type() in filter_type or child.name() in filter_name
+        addSelectChild = selectFilter and isFiltered
```
lower case always except for class declaration please. When we use mixed case, it is because functionnalities from pyqtgraph or Qt do so because they inherit from C++

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -166,14 +166,33 @@ def iter_children(param, childlist=[]):
+        if addSelectChild or addNonSelectChild:
+            childlist.append(child)
         if child.hasChildren():
-            childlist.extend(iter_children_params(child, []))
```
I did this because using list in arguments can cause some issues. I know it seems fine here but I wonder if it can cause some hard to see issues

seb5g reviewed (changes requested):
The utils module also has an iter_children method that gives a list of children names. It would be better if both methods: iter_children and iter_children_param have the same signature and functionality! Could you add the implementation?

Ashwolaa committed:
implementing comments from review

Ashwolaa committed:
- add test

Ashwolaa commented:
> The utils module also has an iter_children method that gives a list of children names. It would be better if both methods: iter_children and iter_children_param have the same signature and functionality! Could you add the implementation?

Is it ok if I merge the two? I think we can just add an extra agument to the iter_children_param to get out all the relevant values that one is interested in? 
If so, do I just add a deprecation signal for the iter_children function which redirects to the iter_children_param method?

seb5g commented:
well no deprecation, it is heavily used in commit_settings in the plugins.
It is meant to quickly check if a given name is within a Parameter structure. So it should only return a list of string not a list of Parameter. However it could have the same filtering (make more sense having similar functions having the same kind of signature). But internally you could just use your new iter_children_param and extract the list of names out of it!

Ashwolaa committed:
need to explicitely give empty string to avoid issue when testing

seb5g commented on src/pymodaq_gui/parameter/utils.py:
```diff
@@ -143,56 +143,61 @@ def compareValuesParameter(param1:Parameter,param2:Parameter,)-> bool:
+    list
+        The list of the children name from the given node.       
+    """
+    return iter_children_params(param, childlist=[], output_type='name', filter_type=(), filter_name=(), select_filter=False)
```
here it should be childlist=childlist

seb5g reviewed (changes requested):
one last change otherwise an initial list would not be populated

Ashwolaa committed:
fix argument transfer between functions
